### ISSUE TITLE
Direct use of torchdynamo.optimizations.analysis fails if you pass in None as an input

### ISSUE THREAD

**Issue by ezyang** [pytorch/torchdynamo#753]
Example

```
diff --git a/fbcode/pytorch/torchdynamo/torchdynamo/optimizations/analysis.py b/fbcode/pytorch/torchdynamo/torchdynamo/optimizations/analysis.py
--- a/fbcode/pytorch/torchdynamo/torchdynamo/optimizations/analysis.py
+++ b/fbcode/pytorch/torchdynamo/torchdynamo/optimizations/analysis.py
@@ -38,8 +38,8 @@
 
     def placeholder(self, target, args, kwargs):
         value = super().placeholder(target, args, kwargs)
-        assert isinstance(value, torch.Tensor)
-        self.input_alias_groups.add(self.tensor_alias_group(value))
+        if isinstance(value, torch.Tensor):
+            self.input_alias_groups.add(self.tensor_alias_group(value))
         return value
 
     def run_node(self, n: torch.fx.Node):
diff --git a/fbcode/torchrec/sparse/jagged_tensor.py b/fbcode/torchrec/sparse/jagged_tensor.py
```

Dynamo will ensure that a None input can never occur but this is not the case for direct use of `compile_fx` api in inductor, for example

**Comment by jansel** [comment]
What does the captured graph do with the None input?  Is it just ignored?

**Comment by ezyang** [comment]
Yeah. What's actually happening is that they're using pytree's to flatten a user defined data structure of Tensors and Nones into a flat structure. It's not convenient to filter out the Nones (because then the pytree unflatten needs to know how to insert Nones at the same spot), but that means they show up in the FX graph.

**Comment by ezyang** [comment]
@voznesenskym mentions that mobile had to write something similar for swizzling Nones around. I deeply feel there is some way we can improve the APIs so this is not necessary.

**Comment by jansel** [comment]
The `List[Tensor] => List[Tensor]` API contract of dynamo backends is useful so that backends don't need to implement support for all python types.

It should be easy to write a shim that turns arbitrary graphs (that take any pytrees supported types) into that normalized `List[Tensor]` format.  We have this same code in AOT Autograd, so just a matter of pulling it out into a `normalize_graph_args()` function.

For mobile @voznesenskym actually was doing the opposite transformation.  De-normalizing graph args to try to match the original function signature.
